base/usr/share/man/man1/more.1
.TH MORE 1 "March, 2026"
.SH NAME
more \- display text by the screenfull.

.SH SYNOPSIS
\fBmore [-r] [-F] [-P \fIstr\fB] [--stay] [--alt] [--scroll] [\fIfile\fB...]\fR

.SH DESCRIPTION
This is an implementation of the common \fBmore\fP utility, which attempts to present its
input one "screenfull" at a time by tracking where it thinks the cursor should be as it
writes to the terminal and waiting for user input when necessary.

.SH OPTIONS

.TP
.B \-r
Assume that basic escape sequences \- those starting with \fI\\033[\fP and ending with a
letter \- do not contribute to the cursor position, ignore them for that purpose, and
emit them raw. This generally allows basic color output to be piped to \fBmore\fP, such
as the output from
.IR roff (1)
or \fIbim \-c\fP.

.TP
.B \-F
If there is only one file, and it fits on the screen without prompting for input,
then quit after printing it.

.TP
.B \-P \fIstr\fP
Instead of showing the name of the file as the prompt, as is the default, show
the specified \fIstr\fP. This is useful if you want to use \fBmore\fP as a manpage
viewer as you can use \fImore -P'man(1)'\fP to set the prompt.

.TP
.B \-\-stay
Instead of quitting when the user asks to advance while at the end of a file,
\fBmore\fP will keep running, like \fBless\fP typically does. This is useful if
you are using \fBmore\fP to interactively page through a file or a manual page.

.TP
.B \-\-alt
Switch to the alternate screen buffer and clear the screen on startup.

.TP
.B \-\-scroll
Disable scrollback. Normally, \fBmore\fP keeps an unlimited scrollback buffer,
so you should set this option if you are piping a large (or unknown) source
into it, or trying to view a very large file.

.TP
.B \-\-help
Shows help text.

.SH SEARCHING
Rudimentary support for searching is provided. To enter a search string, press \fB/\fP.
.PP
Search strings found in processed lines will be highlighted through inverse text.
.PP
Pressing \fBn\fP tries to continue searching forward; pressing \fBN\fP searches backwards.
.PP
Searching is limited to plain text, case\-sensitive searches.

.SH BUGS
\- The \fB-r\fP flag treats all escape sequences as being invisible, and does not handle
ones that are not of the \fBCSI\fP form.
.PP
\- The input prompt does not handle special sequences, but as a consolation it accepts
things like capital \fB'B'\fP as equivalent to \fB'j'\fP or \fBRETURN\fP so you can use the down key,
or \fB'5'\fP as page up.
.PP
\- \fBmore\fP does not update the screen size if the terminal is resized.

.SH SEE ALSO
.BR roff (1),
.BR bim (1)

.SH HOMEPAGE
.PP
.B ToaruOS:
https://toaruos.org/

.SH AUTHOR
This \fImore\fP was written by K. Lange <klange@toaruos.org>.



apps/more.c
/**
 * @brief Print piped input or files one screenful at a time.
 *
 * @copyright
 * This file is part of ToaruOS and is released under the terms
 * of the NCSA / University of Illinois License - see LICENSE.md
 * Copyright (C) 2014-2026 K. Lange
 */
#define _XOPEN_SOURCE 700
#include <errno.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <wchar.h>
#include <unistd.h>
#include <stdlib.h>
#include <locale.h>
#include <termios.h>
#include <signal.h>
#include <sys/signal.h>
#include <sys/ioctl.h>

#include <toaru/decodeutf8.h>

static int term_width = 80;
static int term_height = 25;

static int term_x = 0;
static int term_yish = 1;

static int handle_escapes = 0;
static int stop_asking = 0;
static int exit_if_fits = 0;
static int prompted = 0;
static char * promptstring = NULL;
static int no_quit = 0;
static int use_alt_screen = 0;
static int no_scrollback = 0;
static char * search_string = NULL;
static ssize_t search_size = 0;
static size_t search_avail = 0;
static ssize_t searching = 0;

/* Line builder */
static char * lineBuf = NULL;
static size_t lineLen = 0;
static size_t lineCap = 0;

static void line_add_chr(char c) {
	if (no_scrollback) return;
	if (lineLen == lineCap) {
		lineCap = (lineCap < 5) ? 5 : lineCap * 2;
		lineBuf = realloc(lineBuf, lineCap);
	}
	lineBuf[lineLen++] = c;
}

static void line_add_str(char *s) {
	while (*s) {
		line_add_chr(*s);
		s++;
	}
}

static char ** lines = NULL;
static ssize_t linesLen = 0;
static ssize_t linesCap = 0;

static void lines_add(char * line) {
	if (linesLen == linesCap) {
		linesCap = (linesCap < 5) ? 5 : linesCap * 2;
		lines = realloc(lines, linesCap * sizeof(char*));
	}
	lines[linesLen++] = line;
}

static int to_eight(uint32_t codepoint, char * out) {
	memset(out, 0x00, 7);

	if (codepoint < 0x0080) {
		out[0] = (char)codepoint;
	} else if (codepoint < 0x0800) {
		out[0] = 0xC0 | (codepoint >> 6);
		out[1] = 0x80 | (codepoint & 0x3F);
	} else if (codepoint < 0x10000) {
		out[0] = 0xE0 | (codepoint >> 12);
		out[1] = 0x80 | ((codepoint >> 6) & 0x3F);
		out[2] = 0x80 | (codepoint & 0x3F);
	} else if (codepoint < 0x200000) {
		out[0] = 0xF0 | (codepoint >> 18);
		out[1] = 0x80 | ((codepoint >> 12) & 0x3F);
		out[2] = 0x80 | ((codepoint >> 6) & 0x3F);
		out[3] = 0x80 | ((codepoint) & 0x3F);
	} else if (codepoint < 0x4000000) {
		out[0] = 0xF8 | (codepoint >> 24);
		out[1] = 0x80 | (codepoint >> 18);
		out[2] = 0x80 | ((codepoint >> 12) & 0x3F);
		out[3] = 0x80 | ((codepoint >> 6) & 0x3F);
		out[4] = 0x80 | ((codepoint) & 0x3F);
	} else {
		out[0] = 0xF8 | (codepoint >> 30);
		out[1] = 0x80 | ((codepoint >> 24) & 0x3F);
		out[2] = 0x80 | ((codepoint >> 18) & 0x3F);
		out[3] = 0x80 | ((codepoint >> 12) & 0x3F);
		out[4] = 0x80 | ((codepoint >> 6) & 0x3F);
		out[5] = 0x80 | ((codepoint) & 0x3F);
	}

	return strlen(out);
}

static void char_draw(int c) {
	char tmp[32] = {0};
	if (c == '\t') {
		int count = 8 - (term_x % 8);
		for (int i = 0; i < count; ++i) {
			printf(" ");
			line_add_chr(' ');
		}
	} else if (c < 32 || c == 0x7F) {
		sprintf(tmp, "\033[7m^%c\033[0m", (c < 32) ? ('@' + c) : '?');
	} else if (c > 0x7f && c < 0xa0) {
		sprintf(tmp, "\033[7m<%02x>\033[0m", c);
	} else if (c == 0xa0) {
		sprintf(tmp, "\033[7m \033[0m");
	} else if (c > 127) {
		if (wcwidth(c) >= 1) {
			to_eight(c,tmp);
		} else {
			if (c < 0x10000) {
				sprintf(tmp, "\033[7m[U+%04x]\033[0m", c);
			} else {
				sprintf(tmp, "\033[7m[U+%06x]\033[0m", c);
			}
		}
	} else {
		sprintf(tmp, "%c", c);
	}

	line_add_str(tmp);
	printf("%s", tmp);
}

static int char_width(int c) {
	if (c == '\t') {
		return 8 - (term_x % 8);
	} else if (c < 32 || c == 0x7F) {
		return 2; /* ^@ */
	} else if (c > 0x7f && c < 0xa0) {
		return 4; /* <XX> */
	} else if (c == 0xa0) {
		return 1; /* nbsp */
	} else if (c > 127) {
		int out = wcwidth(c);
		if (out >= 1) return out;
		return (c < 0x10000) ? 8 : 10;
	}
	return 1;
}

static struct termios old;
static void get_initial_termios(void) {
	tcgetattr(STDOUT_FILENO, &old);
}

static void set_unbuffered(void) {
	struct termios new = old;
	new.c_iflag &= (~ICRNL) & (~IXON);
	new.c_lflag &= (~ICANON) & (~ECHO);
	tcsetattr(STDOUT_FILENO, TCSAFLUSH, &new);
}

static void set_buffered(void) {
	tcsetattr(STDOUT_FILENO, TCSAFLUSH, &old);
}

static void clear_line(void) {
	printf("\r\033[K");
	fflush(stdout);
	term_x = 0;
}

static void cleanup(void) {
	if (use_alt_screen) printf("\033[?1007l\033[?1049l");
	printf("\033[?25h");
	set_buffered();
	fflush(stdout);
}

static void quit_cleanly(int sig) {
	cleanup();
	signal(sig, SIG_DFL);
	raise(sig);
}

/**
 * Redraw a line from scrollback, possibly highlighting
 * search results, if found. We try to highlight more
 * than one result in a line, but only if the don't overlap.
 */
static void draw_line(ssize_t line) {
	if (search_string && *search_string) {
		char * stop = NULL;
		for (char * c = lines[line]; *c; ++c) {
			if (!stop && !strncmp(c,search_string,search_size)) {
				printf("\033[7m");
				stop = c + search_size;
			} else if (c == stop) {
				printf("\033[27m");
				stop = NULL;
			}
			fputc(*c, stdout);
		}
		if (stop) printf("\033[27m");
	} else {
		/* No active search, so just print the line */
		printf("%s",lines[line]);
	}
	printf("\033[K\n");
}

/* Quick dumb line editor. */
static int lineedit(char **buffer, size_t *bufsize) {
	size_t c = 0;
	while (1) {
		int i;
		char ch;
		printf("\033[?25h");
		fflush(stdout);
		if (read(STDERR_FILENO, &ch, 1) == 0) {
			i = 0x04; /* Ctrl-D */
		} else {
			i = ch;
		}
		printf("\033[?25l");
		fflush(stdout);

		if (i == 0x08 || i == 127) { /* Backspace */
			if (c) {
				(*buffer)[c--] = '\0';
				printf("\b\033[K");
				fflush(stdout);
			} else {
				return -1;
			}
		} else {
			if (*bufsize <= c + 1) {
				ssize_t nn = *bufsize < 10 ? 10 : (*bufsize * 2);
				char * nbuf = realloc(*buffer, nn);
				*bufsize = nn;
				*buffer = nbuf;
			}

			if (i == 0x04) { /* Ctrl-D which we'll treat as EOF */
				(*buffer)[c] = '\0';
				if (c) return c;
				return -1;
			}

			if (i == '\n' || i == '\r') break; /* unlike getline, we won't include this in the string */
			fputc(i, stdout);
			fflush(stdout);
			(*buffer)[c++] = i;
		}
	}

	(*buffer)[c] = '\0';
	return c;
}

static ssize_t do_search(ssize_t offset, int direction) {
	printf("\r/\033[K");
	fflush(stdout);

	if (direction == 0) {
		if ((search_size = lineedit(&search_string, &search_avail)) == -1) return offset;
	}

	if (search_string && *search_string) {
		if (direction >= 0) {
			/* Search forward. */
			for (ssize_t i = linesLen - term_height - offset + 1 + direction; i < linesLen; ++i) {
				if (i < 0) continue;
				if (strstr(lines[i], search_string)) {
					return linesLen - i - term_height + 1;
				}
			}
			/* No result so set things up so we continue reading input to see
			 * if the search string appears in a line we haven't processed yet. */
			searching = linesLen - term_height - offset;
			stop_asking = 1;
			return -1;
		} else {
			/* Search backwards */
			for (ssize_t i = linesLen - term_height - offset; i >= 0; --i) {
				if (strstr(lines[i], search_string)) {
					return linesLen - i - term_height + 1;
				}
			}
			/* We were already showing the last match; keep the screen
			 * where it is, rather than scrolling it to the top. */
			return offset;
		}
	}

	return offset;
}

static int do_history_mode(int mode, char * title) {
	/* Enter history mode; scroll up one line */
	ssize_t offset = 1;

	if (mode == 2) {
		/* Enter history and scroll to top. */
		offset = linesLen;
	} else if (mode == 3) {
		/* Enter history and "page up" once. */
		offset = term_height;
	} else if (mode == 4) {
		/* Enter history and begin search with prompt. */
		offset = do_search(0, 0);
	} else if (mode == 5) {
		/* Return to history mode from a successful forward search of fresh content */
		searching = 0;
		stop_asking = 0;
		offset = 0;
	} else if (mode == 6) {
		/* Enter history mode from pressing 'n' in non-history mode. (Search next.) */
		offset = do_search(0, 1);
	} else if (mode == 7) {
		/* Enter history mode from pressing 'N' in non-history mode. (Search previous.) */
		offset = do_search(0, -1);
	} else if (mode == 8) {
		/* Enter history mode from failing to find a forward result and hitting EOF. */
		offset = linesLen - searching - term_height;
		stop_asking = 0;
		searching = 0;
	}

	while (1) {
		if (linesLen - term_height + 1 < offset) {
			offset = linesLen - term_height + 1;
		}

		int break_here = 0;
		if (offset < 0) {
			term_yish = term_height + offset;
			offset = 0;
			break_here = 1;
		}

		/* We're going to do this the slow way because actually scrolling is a bit scary. */
		printf("\033[H");
		for (ssize_t line = 1; line < term_height; ++line) {
			ssize_t this = linesLen - term_height - offset + line;
			if (this >= 0 && this < linesLen) draw_line(this);
		}

		if (offset == 0 && break_here) {
			printf("\033[J");
			return 0;
		}

		int attop = (offset == linesLen - term_height + 1);
		printf("\r\033[K\033[7m%s%s\033[0m", title, attop ? " (TOP)" : "");
		printf("\033[?25h");
		fflush(stdout);

		char c;
		read(STDERR_FILENO, &c, 1);
		printf("\033[?25l");
		fflush(stdout);

		switch (c) {
			case '~':
				/* ignore */
				break;
			case 'k':
			case 'A':
				clear_line();
				offset += 1;
				break;
			case ' ':
			case '6':
				clear_line();
				offset -= term_height - 1;
				break;
			case '5':
				clear_line();
				offset += term_height - 1;
				break;
			case 'H':
				clear_line();
				offset = linesLen;
				break;
			case 'F':
			case 'G':
				offset = -1;
				clear_line();
				stop_asking = 1;
				break;
			case 'B':
			case 'j':
			case '\n':
			case '\r':
				clear_line();
				offset -= 1;
				break;
			case '/':
				offset = do_search(offset, 0);
				break;
			case 'n':
				offset = do_search(offset, 1);
				break;
			case 'N':
				offset = do_search(offset, -1);
				break;
			case 'q':
				clear_line();
				return 1;
		}
	}
}

/*
 * This pattern of checking for scrollback, going to history mode
 * with a particular mode, returning if that said to, and then
 * clearing the line to display another prompt is used repeatedly.
 */
#define enter_history_mode(n) { \
	if (no_scrollback) goto _no_scrollback; \
	if (do_history_mode(n, title)) return 1; \
	clear_line(); \
	if (forced) goto _reprint_prompt; \
}

static int next_line(char * title, int forced, int atend) {
	term_yish++;

	/* Whenever we hit "next_line", store the line we were just building */
	if (!forced && !no_scrollback) {
		if (!lineLen) {
			lines_add(strdup(""));
		} else {
			char * c = malloc(lineLen + 1);
			memcpy(c, lineBuf, lineLen);
			c[lineLen] = 0;
			lines_add(c);
		}
		lineLen = 0;
	}

	/* If we were actively searching forward in new data, see
	 * if this newly printed line was a match. */
	if (searching) {
		if (linesLen && strstr(lines[linesLen-1],search_string)) {
			/* Search did match, go back to history mode so it
			 * can redraw the new line with highlighting */
			if (do_history_mode(5,title)) return 1;
			return 0;
		} else if (forced) {
			/* We reached a forced prompt without a search having matched,
			 * which means we reached EOF; go back to history mode as a
			 * failed search. This attempts to restore the scroll
			 * offset from the last search location. */
			if (do_history_mode(8,title)) return 1;
		}
	}

	if (!forced && (term_yish < term_height || stop_asking)) {
		printf("\n");
		term_x = 0;
	} else {
		prompted = 1;
_reprint_prompt:
		printf("%s\033[7m%s%s\033[0m", forced ? "" : "\n", title, atend ? " (END)" : "");
		do {
			printf("\033[?25h");
			fflush(stdout);
			char c;
			read(STDERR_FILENO, &c, 1);
			printf("\033[?25l");
			fflush(stdout);
			switch (c) {
				case '~':
					/* ignore */
					break;
				case '6':
				case ' ':
					term_yish = 1;
					/* fallthrough */
				case 'B': /* Means down generally ends up here */
				case 'j': /* For the vi users. */
				case '\n':
				case '\r':
					clear_line();
					return 0;
				case 'q':
					clear_line();
					return 1;
				case 'F':
				case 'G':
					clear_line();
					stop_asking = 1;
					return 0;
				case 'A':
				case 'k': /* Scroll up one */
					enter_history_mode(1);
					return 0;
				case 'H': /* Scroll to top */
					enter_history_mode(2);
					return 0;
				case '5': /* Page up once */
					enter_history_mode(3);
					return 0;
				case '/': /* Enter search mode, show search prompt */
					enter_history_mode(4);
					return 0;
				case 'n': /* Find next search result */
					enter_history_mode(6);
					return 0;
				case 'N': /* Find previous search result */
					enter_history_mode(7);
					return 0;
				default:
					printf("\r\033[K\033[7munreocgnized command:\033[0m %c", c);
					fflush(stdout);
					break;
			}

			continue;
_no_scrollback:
			clear_line();
			printf("\r\033[K\033[7mScrollback is disabled.\033[0m");
			fflush(stdout);
		} while (1);
	}

	return 0;
}

static int handle_one(int code, char * name) {
	int width = char_width(code);
	if (term_x + width > term_width) {
		if (next_line(promptstring ? promptstring : name, 0, 0)) return 1;
	}
	char_draw(code);
	term_x += width;
	return 0;
}

static int do_file(char * name, FILE * f, int opti, int argc) {
	if (linesLen) linesLen = 0;
	if (lineLen) lineLen = 0;

	if (!f) {
		printf("%s: %s\n", name, strerror(errno));
		return next_line("Press RETURN to continue, q to exit.", 1, 0);
	}
	int is_escaped = 0;
	int maybe_escaped = 0;

	term_x = 0;
	searching = 0;
	stop_asking = 0;

	uint32_t code, state = 0;
	while (!feof(f)) {
		int c = fgetc(f);
		if (c < 0) break;
		if (!decode(&state, &code, c)) {
			if (code == '\n') {
				if (next_line(promptstring ? promptstring : name, 0, 0)) return 1;
			} else if (is_escaped) {
				if (code >= 'A' && code <= 'z') {
					is_escaped = 0;
				}
				char tmp[8] = {0};
				to_eight(code,tmp);
				printf("%s", tmp);
				line_add_str(tmp);
			} else if (maybe_escaped) {
				if (code == '[') {
					maybe_escaped = 0;
					is_escaped = 1;
					printf("\033[");
					line_add_str("\033[");
				} else {
					handle_one('\033',name);
					handle_one(code,name);
				}
			} else if (handle_escapes) {
				if (code == '\033') {
					maybe_escaped = 1;
				} else {
					handle_one(code,name);
				}
			} else {
				handle_one(code,name);
			}
		} else if (state == UTF8_REJECT) {
			state = 0;
		}
	}

	/* If we never prompted, the file fit on the screen. If this is the last (only) file,
	 * and we had the -F option, we can skip printing the forced END prompt. */
	if (!prompted && exit_if_fits && opti + 1 >= argc) return 0;

	do {
		stop_asking = 0;
		if (next_line(promptstring ? promptstring : name, 1, 1)) return 1;
	} while (no_quit);
	return 0;
}

static int usage(char * argv[]) {
#define X_S "\033[3m"
#define X_E "\033[0m"
	fprintf(stderr,
		"usage: %s [-r] [-F] [-P " X_S "str" X_E "] [" X_S "file" X_E "...]\n"
		"\n"
		"Print files one screenful at a time. If no files are provided,\n"
		"attempts to read from stdin only if it is not a terminal.\n"
		"\n"
		"Options:\n"
		"\n"
		" -r       " X_S "Try to parse some escape sequences to support, eg., color." X_E "\n"
		"          " X_S "Only 'm' sequences are likely to work; this option is meant" X_E "\n"
		"          " X_S "for things like manpages or 'bim -c'." X_E "\n"
		" -F       " X_S "If the file fits on one screen, exit after printing." X_E "\n"
		"          " X_S "This option only works if there is only one file." X_E "\n"
		" -P " X_S "str   Instead of showing file names as the command prompt, use the" X_E "\n"
		"          " X_S "provided string, eg. '-P man' will display the prompt 'man'." X_E "\n"
		" --stay   " X_S "Normally, more will quit when asked to advance when at the end" X_E "\n"
		"          " X_S "of a file, but with this option it will stay running until a" X_E "\n"
		"          " X_S "more explicit quit command (eg. 'q') is issued." X_E "\n"
		" --alt    " X_S "Switch to the alternate buffer and clear the screen on startup." X_E "\n"
		" --scroll " X_S "Disable scrollback." X_E "\n"
		" --help   " X_S "Show this help text." X_E "\n"
		"\n"
		"Command input is received on stderr. The following commands are accepted when\n"
		"prompted during viewing of a file:\n"
		"\n"
		" RETURN   " X_S "Proceed to the next line, or the next file if at the end." X_E "\n"
		"          " X_S "'B' and 'j' are also accepted as aliases." X_E "\n"
		" SPACE    " X_S "Proceed to the next screenful of text." X_E "\n"
		" G        " X_S "Continue outputting this file until reaching the end." X_E "\n"
		" q        " X_S "Quit immediately, ignoring all other files." X_E "\n"
		"\n", argv[0]);
	return 1;
}

int main(int argc, char * argv[]) {
#ifdef __APPLE__
	/* TODO figure out a better way to do this; maybe just LC_CTYPE? */
	setlocale(LC_ALL, "en_US.UTF-8");
#else
	setlocale(LC_ALL, "");
#endif

	int opt;

	while ((opt = getopt(argc, argv, "rFP:-:")) != -1) {
		switch (opt) {
			case 'r':
				handle_escapes = 1;
				break;
			case 'F':
				exit_if_fits = 1;
				break;
			case 'P':
				promptstring = optarg;
				break;
			case '-':
				if (!strcmp(optarg,"help")) {
					usage(argv);
					return 0;
				}
				if (!strcmp(optarg,"stay")) {
					no_quit = 1;
					break;
				}
				if (!strcmp(optarg,"alt")) {
					use_alt_screen = 1;
					break;
				}
				if (!strcmp(optarg,"scroll")) {
					no_scrollback = 1;
					break;
				}
				fprintf(stderr, "%s: '--%s' is not a recognized long option\n", argv[0], optarg);
				/* fallthrough */
			case '?':
				return usage(argv);
		}
	}

	if (optind == argc && isatty(STDIN_FILENO)) {
		fprintf(stderr, "%s: stdin is a TTY and no file names were provided.\n", argv[0]);
		return 1;
	}

	if (!isatty(STDOUT_FILENO)) {
		fprintf(stderr, "%s: This implementation refuses to write to a non-terminal.\n", argv[0]);
		return 1;
	}

	struct winsize w;
	ioctl(STDOUT_FILENO, TIOCGWINSZ, &w);
	term_width  = w.ws_col;
	term_height = w.ws_row;
	get_initial_termios();
	set_unbuffered();
	printf("\033[?25l");

	signal(SIGINT, quit_cleanly);
	signal(SIGQUIT, quit_cleanly);

	if (use_alt_screen) printf("\033[?1049h\033[?1007h\033[H\033[2J");

	if (optind == argc) {
		do_file("stdin",stdin, optind, argc);
	}

	for (; optind < argc; optind++) {
		FILE * f = fopen(argv[optind], "r");
		if (do_file(argv[optind], f, optind, argc)) break;
		if (f) fclose(f);
	}

	cleanup();
	return 0;
}
